### PR TITLE
Have NuGetRestorer only respond to PackageRefs capability

### PR DESCRIPTION
**Pull request by jmarolf**
**Customer scenario**

Customer is getting errors from the project system attempting to run a restore on a shared project.

**Bugs this fixes:** 

Fixes https://github.com/dotnet/project-system/issues/2231

**Workarounds, if any**

None

**Risk**

Code is hard to unit test.  Manually buddy testing is required to merge this.

**Performance impact**

This code has the potential to change the perf for nuget restores so it should be validated before checkin

### CODE REVIEW

**Comment by jmarolf** [comment]
@dotnet/project-system please review

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs [hunk 1]
```diff
@@ -67,14 +67,17 @@ internal Task OnProjectFactoryCompletedAsync()
                 ruleNames: _targetFrameworkWatchedRules,
                 initialDataAsNew: false, // only reset on subsequent changes
                 suppressVersionOnlyUpdates: true);
-
             return Task.CompletedTask;
         }
 
+        public Task UnloadAsync() => Task.CompletedTask;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs]: What I am I missing here? You haven't unsubscribed.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs [hunk 2]
```diff
@@ -16,7 +16,9 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.NuGet
 {
     using TIdentityDictionary = IImmutableDictionary<NamedIdentity, IComparable>;
 
-    internal class NuGetRestorer : OnceInitializedOnceDisposedAsync
+    [Export(ExportContractNames.Scopes.ConfiguredProject, typeof(IProjectDynamicLoadComponent))]
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs]: This used to be an "unconfigured project service" - it now lives in the configured project scope, which means that one is created for every single configuration. You'll want this to be UnconfiguredProject.

**Comment by jmarolf** [comment]
I can't get anything bad to happen in shared projects.  Semantic classification and errors still work in d15rel with shared projects

**Comment by jmarolf** [comment]
@davkean ping

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs [hunk 3]
```diff
@@ -67,14 +67,17 @@ internal Task OnProjectFactoryCompletedAsync()
                 ruleNames: _targetFrameworkWatchedRules,
                 initialDataAsNew: false, // only reset on subsequent changes
                 suppressVersionOnlyUpdates: true);
-
             return Task.CompletedTask;
         }
 
+        public Task UnloadAsync() => DisposeAsync();
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs]: No, Unload != Dispose. Unload is "temporarily shutdown". 

Easiest way to test this is to add Directory.Build.targets in the same directory and continually comment/uncomment and Save the file.

``` XML
 <ItemGroup>
    <ProjectCapability Remove="PackageReferences"/>
  </ItemGroup>
```

This should cause Load/Unload to be called over and over again.

**Comment by jmarolf** [comment]
@davkean ping

**Comment by davkean** [comment]
@jmarolf The change you've made doesn't resolve my concern. Can you sync with me offline and we can discuss?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs [hunk 4]
```diff
@@ -56,8 +59,10 @@ internal class NuGetRestorer : OnceInitializedOnceDisposedAsync
             _activeConfiguredProjectsProvider = activeConfiguredProjectsProvider;
         }
 
+        // In order for the initial project restore to work, we need to wait for the project to be configured
+        // hence startAfter: ProjectLoadCheckpoint.ProjectFactoryCompleted.
         [ProjectAutoLoad(startAfter: ProjectLoadCheckpoint.ProjectFactoryCompleted)]
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs]: This needs to be removed.

**Comment by shyamnamboodiripad** [comment]
@jmarolf I have retargeted this PR to master since the dev15.6 branch is being deleted (master is the new branch for anything post-15.3).

cc @dotnet/project-system @srivatsn

**Comment by davkean** [comment]
I have a pending change that picks up this PR.

**Pull request closed without merging**